wgengine/magicsock/magicsock.go

```diff
@@ -4057,12 +4091,11 @@ type endpoint struct {
 	fakeWGAddr   netip.AddrPort // the UDP address we tell wireguard-go we're using
 	nodeAddr     netip.Addr     // the node's first tailscale address; used for logging & wireguard rate-limiting (Issue 6686)
 
+	disco atomic.Pointer[endpointDisco] // if the peer supports disco, the key and short string
+
 	// mu protects all following fields.
 	mu sync.Mutex // Lock ordering: Conn.mu, then endpoint.mu
```

Comment:
ugh, so some fields are behind a mutex and some are an atomic?

Reply:
I know. I looked at the mutex strategy and due to the way these two fields are accessed moving over to using the mutex properly would have been painful and require restructuring more implementations. The mutex is used inconsistently today.

I'm not particularly happy with it either, but the root cause is at the higher level organization - that requires us to sit down and review the design in order to address it.

Reply:
Maybe I'm missing it, but I can't see from the comments on #7821 what the actual data race was.

Which code path was using these fields without the proper locks being held?

Reply:
Conn.upsertEndpoint, Conn.deleteEndpoint, Conn.handleDiscoMessage, Conn.unambiguousNodeKeyOfPingLocked, usage of ep.shortString in a number of logf calls, and endpoint.String.

Of all of these the one most tricky to address is endpoint.String has it has no clear lock bounds and is a generic interface that escapes into a foreign code path, it may not be safe to introduce a lock there.

I could write a version of this that uses the mutex everywhere, but drops the disco key from endpoint.String?